[WIP] Testing and adapting for php7 if necessary. (Including an increase in Code Coverage)

WillSkates opened the pull request.

garemoko commented:
I'm assuming this PR is a work in progress. See https://github.com/RusticiSoftware/TinCanPHP/issues/46 which may be helpful.

WillSkates commented:
Yep you're correct thanks for the link. I've changed the title to reflect.

WillSkates commented:
Ready for merge.

garemoko commented:
@johnpbloch as you raised issue #46, would you like to review this PR?

WillSkates commented:
@garemoko Are you happy with this PR?

garemoko commented:
@WillSkates thanks for the poke. I was hoping that @johnpbloch might be set up to test this. 

Nevertheless, I'll add this to me to-do list to test this branch with php 5.4 which I have installed now, and then see if I can upgrade to php 7 and test with that.

johnpbloch commented:
Sorry, I was travelling at the time, and this fell off my radar since then. I'm checking the tests right now. Regarding a code review, I'm very much not familiar with the codebase, so I'll leave that to someone who is better qualified to review it.

johnpbloch commented:
Well, the PHP7 tests aren't passing for me locally, but neither are the tests in master, so this is probably something wrong with my environment. I will try again later.

WillSkates commented:
@johnpbloch No problem, thank you for coming back so quickly after the poke. Could you please paste your output? You will need to provide some configuration before you run them i.e. https://github.com/WillSkates/TinCanPHP/blob/master/tests/Config.php.travis-ci 

@garemoko It might be worth looking at using a VM or a Docker image to run the test suite in.

johnpbloch commented:
Output of tests on master:

```sh
TinCanPHP (master=) $ /c/PHP56/php vendor/phpunit/phpunit/phpunit tests
PHPUnit 5.1.4 by Sebastian Bergmann and contributors.

..............................................................  62 / 189 ( 32%)
.........................F.F............F...F................. 124 / 189 ( 65%)
.......................................E......FFFF............ 186 / 189 ( 98%)
...                                                            189 / 189 (100%)

Time: 11.73 seconds, Memory: 7.00Mb

There was 1 error:

1) StatementTest::testVerifyIncorrectPubKey
openssl_pkey_get_details() expects parameter 1 to be resource, boolean given

C:\Users\John\Projects\TinCanPHP\tests\StatementTest.php:811

--

There were 8 failures:

1) RemoteLRSTest::testSaveStatementsWithAttachments
success
Failed asserting that false is true.

C:\Users\John\Projects\TinCanPHP\tests\RemoteLRSTest.php:171

2) RemoteLRSTest::testRetrieveStatementWithAttachments
attachment content
Failed asserting that null is identical to '{"foo":"bar"}'.

C:\Users\John\Projects\TinCanPHP\tests\RemoteLRSTest.php:242

3) RemoteLRSTest::testRetrieveActivity
retrieved activity
'{"error":true,"success":false...ce"}]}' does not match expected type "object".

C:\Users\John\Projects\TinCanPHP\tests\RemoteLRSTest.php:481

4) RemoteLRSTest::testRetrievePerson
retrieved person
Failed asserting that two objects are equal.
--- Expected
+++ Actual
@@ @@
 TinCan\Person Object (
     'objectType' => 'Person'
     'name' => Array (
-        0 => 'Example User'
     )
     'mbox' => Array (
-        0 => 'mailto:tincanphp@tincanapi.co...person'
     )
-    'mbox_sha1sum' => null
-    'openid' => null
-    'account' => null
+    'mbox_sha1sum' => Array ()
+    'openid' => Array ()
+    'account' => Array ()
 )

C:\Users\John\Projects\TinCanPHP\tests\RemoteLRSTest.php:541

5) StatementVariationsTest::testAttachmentsString
successful request
Failed asserting that false is true.

C:\Users\John\Projects\TinCanPHP\tests\StatementVariationsTest.php:116

6) StatementVariationsTest::testAttachmentsBinaryFile
Failed asserting that false is true.

C:\Users\John\Projects\TinCanPHP\tests\StatementVariationsTest.php:152

7) StatementVariationsTest::testSignedAndVerified
verify signature
Failed asserting that false is true.

C:\Users\John\Projects\TinCanPHP\tests\StatementVariationsTest.php:186

8) StatementVariationsTest::testSignedAndVerifiedX5c
verify signature
Failed asserting that false is true.

C:\Users\John\Projects\TinCanPHP\tests\StatementVariationsTest.php:220

FAILURES!
Tests: 189, Assertions: 711, Errors: 1, Failures: 8.


Non-exception error occurred:

2: openssl_x509_read(): supplied parameter cannot be coerced into an X509 certificate!
C:\Users\John\Projects\TinCanPHP\src\Statement.php (254)
```

Output from the php7 branch:

```sh
TinCanPHP (php7=) $ vendor/bin/phpunit tests
PHPUnit 5.1.4 by Sebastian Bergmann and contributors.

..............................................................  62 / 191 ( 32%)
...........................F.F............F...F............... 124 / 191 ( 64%)
.........................................E......FFFF.......... 186 / 191 ( 97%)
.....                                                          191 / 191 (100%)

Time: 17.39 seconds, Memory: 6.00Mb

There was 1 error:

1) StatementTest::testVerifyIncorrectPubKey
openssl_pkey_get_details() expects parameter 1 to be resource, boolean given

C:\Users\John\Projects\TinCanPHP\tests\StatementTest.php:828

--

There were 8 failures:

1) RemoteLRSTest::testSaveStatementsWithAttachments
success
Failed asserting that false is true.

C:\Users\John\Projects\TinCanPHP\tests\RemoteLRSTest.php:171

2) RemoteLRSTest::testRetrieveStatementWithAttachments
attachment content
Failed asserting that null is identical to '{"foo":"bar"}'.

C:\Users\John\Projects\TinCanPHP\tests\RemoteLRSTest.php:242

3) RemoteLRSTest::testRetrieveActivity
retrieved activity
'{"error":true,"success":false...ce"}]}' does not match expected type "object".

C:\Users\John\Projects\TinCanPHP\tests\RemoteLRSTest.php:481

4) RemoteLRSTest::testRetrievePerson
retrieved person
Failed asserting that two objects are equal.
--- Expected
+++ Actual
@@ @@
 TinCan\Person Object (
     'objectType' => 'Person'
     'name' => Array (
-        0 => 'Example User'
     )
     'mbox' => Array (
-        0 => 'mailto:tincanphp@tincanapi.co...person'
     )
-    'mbox_sha1sum' => null
-    'openid' => null
-    'account' => null
+    'mbox_sha1sum' => Array ()
+    'openid' => Array ()
+    'account' => Array ()
 )

C:\Users\John\Projects\TinCanPHP\tests\RemoteLRSTest.php:541

5) StatementVariationsTest::testAttachmentsString
successful request
Failed asserting that false is true.

C:\Users\John\Projects\TinCanPHP\tests\StatementVariationsTest.php:116

6) StatementVariationsTest::testAttachmentsBinaryFile
Failed asserting that false is true.

C:\Users\John\Projects\TinCanPHP\tests\StatementVariationsTest.php:152

7) StatementVariationsTest::testSignedAndVerified
verify signature
Failed asserting that false is true.

C:\Users\John\Projects\TinCanPHP\tests\StatementVariationsTest.php:186

8) StatementVariationsTest::testSignedAndVerifiedX5c
verify signature
Failed asserting that false is true.

C:\Users\John\Projects\TinCanPHP\tests\StatementVariationsTest.php:220

FAILURES!
Tests: 191, Assertions: 700, Errors: 1, Failures: 8.


Non-exception error occurred:

2: openssl_x509_read(): supplied parameter cannot be coerced into an X509 certificate!
C:\Users\John\Projects\TinCanPHP\src\Statement.php (254)
```

Contents of test/Config.php:

```php
<?php

$LRSs = [
    [
        'endpoint' => 'http://localhost:8000/data/xAPI/',
        'username' => '402da968f21dee8300df6cc7ded89dea5e3665e8',
        'password' => '142a594f5dbcdf8d5d693b27d92d46925a08ce95',
        'version'  => '1.0.1'
    ]
];
$KEYs = [
    'public' => __DIR__ . '/keys/travis/cacert.pem',
    'private' => __DIR__ . '/keys/travis/privkey.pem',
    'password' => 'travis',
];
```

The LRS credentials are from the instance of LearningLocker I have set up locally for testing our own LRS systems and I started this file from scratch and added the credentials directly from the web interface before running the tests, so I know they're correct. The learninglocker version is a bit out of date though, which may be contributing to this. I'll test this again with a more recent version of LL.

garemoko commented:
@johnpbloch I imagine that these tests were originally designed against SCORM Cloud. Aside from actual bugs in whichever Learning Locker version you're using, there may be legitimate style issues that cause failed tests. For example on test 4, Learning Locker has returned empty arrays where the test expected the properties to be null. Either are arguably valid by the spec. 

Ideally the tests should be made to work with any conformant LRS, but that's out of scope of this PR.

garemoko commented:
Note: you can get a free SCORM Cloud account for testing here: https://cloud.scorm.com/sc/guest/SignUpForm

brianjmiller commented:
You can also trigger a Travis build yourself on your own fork through their site which will run the tests as I will ultimately. Just beware that simultaneous test runs against the same Cloud account can be problematic, unfortunately the tests aren't safe that way.

WillSkates commented:
Thank you for the input everyone. Is there a reason why the Travis environment wont guarantee that the build should work in any environment assuming that it is configured correctly?

brianjmiller commented:
If I understand what you are asking, not really, short of a bug in Travis which happens from time to time. That consistency is one of the big reasons to rely on Travis.

WillSkates commented:
@brianjmiller Yeah sorry I didn't really write that well. Are you happy with the PR in general?

WillSkates commented:
Accidentally slipped and clicked close rather than comment :p.

garemoko commented on .gitignore:
```diff
@@ -22,4 +22,4 @@ todo.md
 .buildpath
 .project
-.settings
+.settings
```
Please avoid whitespace changes in files that haven't otherwise changed in order to avoid wrongly updating the "last updated" date.

brianjmiller replied:
This hasn't been addressed.

garemoko commented on src/FromJSONTrait.php:
```diff
@@ -27,9 +27,38 @@ public static function fromJSON($jsonStr) {
+         * recommended version of php is changed to 5.5. Especially seeing as php 5.4
+         * is now no longer being supported as of me writing this.
+         * @see http://php.net/supported-versions.php
+         */
```
Can we remove this comment and instead raise an issue here in Github? 

In regards to PHP versions, I suggest that we support PHP versions at least in line with what the oldest of what latest Moodle version and latest Wordpress versions support to ensure that this library continues to be compatible with both.

WillSkates replied:
Yep the comment can go. With regards to versions upping these to 5.6.* should be fine given that 5.5 won't be receiving security updates in 5 months time.

0.12.0 will still support 5.5 and 5.4.

garemoko replied:
My main concern here is getting ahead of Moodle's require php version because it means that we have to require a higher PHP version for the Moodle plugins than for Moodle itself, and that's even assuming that the latest version of Moodle is being used. 

That said, the Moodle Logstore plugin already requires PHP 5.5 so perhaps that's not a good argument.

WillSkates replied:
I honestly think that doing this is fine seeing as the older versions of the library and the plugin will still be available and will work fine. To make it clearer the documentation can also be changed to show which version of the library is compatible with which php version.

garemoko commented on src/FromJSONTrait.php:
```diff
@@ -27,9 +27,38 @@ public static function fromJSON($jsonStr) {
+
+         if (isset($errors[$code])) {
+           return $errors[$code];
+         } else {
```
Please start elses on a new line.

garemoko commented on src/FromJSONTrait.php:
```diff
@@ -27,9 +27,38 @@ public static function fromJSON($jsonStr) {
     }
+
+    private static function provideJsonErrorMessageForCode($code)
+    {
```
Please put opening braces for functions on the same line.

brianjmiller replied:
Not addressed.

garemoko commented on src/JsonParseErrorException.php:
```diff
@@ -0,0 +1,57 @@
+    private $jsonErrorMessage;
+
+    public function __construct($malformedValue, $jsonErrorNumber, $jsonErrorMessage, Exception $previous = null)
+    {
```
Please put opening braces for functions on the same line.

garemoko replied:
(Also applies below - I'll not flag up every instance)

WillSkates replied:
Could you please link me to the standards guide that you would like me to follow? This code should be PSR1 and 2 compliant although I must admit I haven't gone through it and checked.

garemoko replied:
I'm not aware of a particular code standard we're following, it's just about consistency with the code in the rest of the library. @brianjmiller may have a better answer here.

WillSkates replied:
Cool @brianjmiller it would be useful if it's one of [these](https://github.com/squizlabs/PHP_CodeSniffer/tree/master/CodeSniffer/Standards) so that we can automate the process of finding inconsistencies.

garemoko commented on src/JsonParseErrorException.php:
```diff
@@ -0,0 +1,57 @@
+class JsonParseErrorException extends Exception
+{
+
+    private static $msg = 'The provided json value "%s" couldn\'t be parsed. json_decode() provided the following error message: "%s" with the error code %d.';
```
JSON instead of json.

WillSkates replied:
Agreed

garemoko commented on tests/JsonParseErrorExceptionTest.php:
```diff
@@ -0,0 +1,54 @@
+        );
+    }
+
+    public function testThatWeCanGetOurInformationBackOut()
```
It'd be good to use the same style/tone of function name as used in other tests. 

Possibly this should be three separate tests; I can't think of a good collective name.

WillSkates replied:
I think the main distinction is that I haven't used a test to show that we can create an object successfully. The code in setUp() should achieve the same thing.

garemoko commented on tests/JsonParseErrorExceptionTest.php:
```diff
@@ -0,0 +1,54 @@
+        $this->assertEquals($this->jsonErrorMessage, $this->exception->jsonErrorMessage());
+    }
+
+    public function testThatWeGetAMeaningfulErrorMessage()
```
testGetMessage

WillSkates replied:
That doesn't seem descriptive enough as "Get" has lots of definitions (around 40-50). Although I do think that my method name here is a little too descriptive. testThatErrorsAreMeaningful()?

garemoko replied:
Again, it's about consistency with the rest of the library. Take a look at the function names in the other tests.

WillSkates replied:
Its fine to be inconsistent here given that there is a demonstrated improvement. I'm not breaking consistency for the sake of it.

The potential impact is that the code will look alien in comparison to what is already there. Meaning that at first glance it is more _difficult to understand_ because it doesn't follow a pattern. Given that this is (in my view) _easier to understand_ it solves the same problem it creates by not conforming.

garemoko replied:
@WillSkates i'll leave this one to @brianjmiller :-)

garemoko commented on tests/StatementTest.php:
```diff
@@ -39,24 +39,21 @@ public function testInstantiation() {
         $this->setExpectedException(
-            'InvalidArgumentException',
-            'Invalid JSON: ' . JSON_ERROR_SYNTAX
+            'TinCan\JsonParseErrorException'
```
Is this testing that the right error is returned in the right context? It doesn't look like it is since no parameter is passed?

garemoko replied:
This applies to all the tests of this type, not just statementtest.

WillSkates replied:
We would effectively be testing how json_encode() behaves under certain types of input which is already done for us [e.g](https://github.com/php/php-src/tree/master/ext/json/tests).

All we need to test is whether our exception returns a message that is formatted how we want that includes output from json_last_error(). That is done in JsonParseErrorExceptionTest::testThatWeGetAMeaningfulErrorMessage().

garemoko commented:
@WillSkates Thanks again for this and sorry it took me a little time to get to. Thanks especially for adding tests. 

I've added some comments on the code for you to take a look at, mainly coding style things to match the style of the rest of the repo. 

I also tested this version with the Moodle logstore plugin on MAMP running php5.6.2, which is what I happened to have to hand. It works fine. 

@johnpbloch did you get to a place where you are happy with this PR?

WillSkates commented:
@garemoko Sorry by the way as I haven't said thanks for taking the time to look at this pr. Thanks :D.

garemoko commented:
@WillSkates no problem!

WillSkates commented:
@brianjmiller If you have ten minutes we need your resolution on some method naming. I also would like to know what coding standard and/or style guidelines I should be following.

WillSkates commented:
@brianjmiller bump as we are still waiting on your input.

WillSkates commented:
@garemoko As this has been sat here for two weeks will you merge this if I make the whole thing PSR 1 and 2 compliant?

garemoko commented:
@WillSkates I can't merge; only @brianjmiller has that power.

WillSkates commented:
@garemoko Ah ok. Is this being ignored for any reason?

WillSkates commented:
Any news?

WillSkates commented:
@garemoko Thank you for our conversations about the code above I genuinely have enjoyed them. 

I need to get out of this situation where the only person with the ability to merge is uncommunicative for nearly a month. Would you like to fork or rework this project so that we could be more active?

I'm guessing the license is actually void given that [the owners information is missing](https://github.com/RusticiSoftware/TinCanPHP/blob/master/LICENSE#L190).

Cheers,
Will

WillSkates commented:
@garemoko

garemoko commented:
@WillSkates Heh. I'm not going to fork this project myself, but I do wonder if @brianjmiller might be open to adding collaborators who can help with review and merge of code. I'm not the right person for that, but I can think of some other options. I'll mail Brian.

brianjmiller commented:
Frankly I was keeping an eye on the PR and knew that Andrew had the review in hand, I thought the email notifications were still fixes. I'll try to make sometime within the next week or so to review this and we'll get it merged in. (Thanks for the out of band email.)

I can say I'll agree with @garemoko if the code itself doesn't follow the conventions of the rest of the library, whether or not they follow any community style guidelines, I'll reject it favoring consistency, though I do recall trying to meet the PSR*s available at the time we originally developed it so meeting them should make it close.

brianjmiller commented:
Adding collaborators is above my pay grade. We've had so few community additions that I'm not sure it would serve a lot of purpose, sorry this one has lingered longer than you'd like, but a month on an OSS library just isn't a long time especially when targeting a new runtime version, and presumably you can run your code in the meantime.

brianjmiller commented on composer.json:
```diff
@@ -24,8 +24,9 @@
-        "php": "~5.4",
+        "php": ">=5.4",
         "ext-openssl": "*",
+        "ext-curl": "*",
```
Why are we adding curl as a dependency?

WillSkates replied:
Purely for consistency across my environment, travis and @johnpblochs testing environment. Where the current stream_context* did, didn't and didn't work respectively.

At the time we were testing against 7.0-RC5 and 7.0-RC6 I believe. Now that we have a stable release I'll add reverting the CURL implementation to the list.

brianjmiller commented:
It doesn't seem like the original comments were addressed? I'd rather not double review what someone else has already pointed out as problems. Need to address the style consistency items, at least up to the point at which Andrew has noted them.

WillSkates commented:
@brianjmiller Thank you for coming back and taking a look. You don't need to re review the code. 

Here's the things that we needed you to respond to:

1. Please provide a set of coding style guidelines or say yes to making the lib PSR1 + PSR2 compliant.
2. Is it ok to have it so that the latest version of the library will require PHP versions that are currently being supported (i.e. >5.5) given that users on the older version of the lib will still be able to use the older version?
3. Is it ok to break convention in "tests/JsonParseErrorExceptionTest.php" given that the method names (IMO) make it easier to derive what the test is doing?
4. Do you agree that we don't need to test how json_last_error() will behave given that this is already done for us in the module itself?

The original comments have not been addressed because we require your resolution on the above 4 points. I planned to make the other smaller changes i.e. "Json->JSON" as I was going with the bigger ones.

WillSkates commented:
Any news on the above ^.

brianjmiller commented:
@WillSkates 

1) Is still the same, the guidelines are to follow the established convention of the library, in particular maintaining consistency especially at the file level. I don't plan to entertain wide changes to the library as a whole purely for subjective style considerations, at least not in the same PR/release where we add another PHP engine.

2) I think it is fine if we are requiring a minimum version supported by the PHP core community, having said that I don't see a reason to break support unless there is an actual gain. I've not looked specifically at why we need to break support, I will do so later today.

3) I tend to agree with @garemoko on this, the testing names as they are right now point very specifically to the API calls that are being made so that you can know from that test name which API is tested and gauge at least to an extent the coverage you have. For instance we don't care that the message is meaningful, we care that the method/property returns what we expect, whether or not that is non-sensical (or meaningful) isn't really being tested.

4) I'm not sure I even understand this yet, but I think I agree if by the "module" we mean core. Having said that I'm not sure why these changes are even being introduced in this PR and I have concerns that we are breaking backwards compatibility in general. Changing the type of the exception returned in the `fromJSON` calls breaks backwards compatibility and necessitates a major version bump of the library just to make debugging something pretty trivial to debug slightly easier and is only necessary for *very* odd cases of someone generating the JSON manually as the library itself *should* be creating conformant JSON and an LRS *must* be generating conformant objects for itself to be conformant, so I don't see what the win here is for introducing breaking changes for a more specific exception. Even if it is a significant win I don't know why it is combined into a PR targeting the bump to PHP 7. They should be isolated change sets.

Really before going much further this PR needs to minimally comply with the style consistency request, needs to be squashed (no reason for this to be 52 commits), need to revert the inclusion of curl unless for some reason PHP 7 doesn't support/provide the native interface, and need to determine whether or not we can avoid any backwards incompatible changes and separate any unnecessary breakage. And more than likely we need to make sure that this PR can fit neatly on top of the other open PRs as I'm likely to be merging them while I'm working on this library as they've been sitting out much longer.

brianjmiller commented:
@WillSkates now had a chance to understand the JSON changes. 

* I think this necessitates a major version bump, but I don't think that is going to be avoidable with PHP 7 which we do need to support, so the time might as well be now. I think it is fine to drop <5.5 support to simplify the handling here.
* We don't need to test that `json_last_error` is sane, **but** we do need to test that our public interface is sane which is effectively the same thing. And we want to make sure to do that so that we don't get caught when they change the interface again.
* I'm not sure the `composer.json` update is the best option, that is allowing anything >= 5.4, I think we are better off specifying as `~5.5 || ^7.0` (see https://github.com/RusticiSoftware/TinCanPHP/pull/50/files).

I think the rest of the comments still apply.

brianjmiller commented:
FWIW, I will be out of the office starting tomorrow until 03/02 so any further review is not likely to occur in that time period.

WillSkates commented:
Thanks for the rest of the updates. I'm going to be tied up for the next week or two as well so I'll do as I can.

WillSkates commented:
Apologies for not getting back to this sooner as I've been unable to dedicate any time to this except reverting the libcurl change. I will push once I've played around with the environment to reproduce the errors seen here : https://travis-ci.org/RusticiSoftware/TinCanPHP/builds/98109869 so that I can fix them as well.

WillSkates commented:
@brianjmiller @garemoko I think those last 4 commits cover just about everything in our notes.

I've messed around trying to remove my changes to .gitignore with little success. Could you please advise on how to do this?

I've reworked the format of the JSONParseErrorException message to be a lot less wordy. The old message format seemed far too long for people to read. We'd also need to check if it still made sense based upon what content was in it.

I haven't added tests for json_last_error() message behavior based on malformed input. What we have here tests that we are wrapping the output of json_last_error() in an exception that someone can catch (JSONParseErrorException). This is all that exception does. We got caught out cause we were double checking something that didn't need to be checked again.

Cheers,
Will

WillSkates commented:
Is what we have here ok? I will squash as soon as @brianjmiller confirms it is ready to merge.

garemoko commented:
@WillSkates do you mind taking a quick look at https://github.com/RusticiSoftware/TinCanPHP/pull/47/files#r50035452 ?

WillSkates commented:
Hi @garemoko yeah I wasn't able to find a way to fix that. Can you advise?

garemoko commented:
I would just copy the file from the master branch and overwrite your changed file.

WillSkates commented:
I kinda recklessly found a fix and squashed.

garemoko commented on src/Statement.php:
```diff
@@ -251,6 +251,7 @@ public function verify($options = array()) {
         }
         else if (isset($header['x5c'])) {
             $cert = "-----BEGIN CERTIFICATE-----\r\n" . chunk_split($header['x5c'][0], 64, "\r\n") . "-----END CERTIFICATE-----\r\n";
+
```
Not sure if this was here last time I reviewed; apologies for missing it if it was. 

It doesn't seem like I good idea to edit a file and bump the last modified date just to add a new line.

garemoko commented:
Aside from https://github.com/RusticiSoftware/TinCanPHP/pull/47#discussion_r57134989 this looks good to me. 

I don't fully understand why the error stuff has to change or what the new code is doing, but I'm trusting there's a good reason for it.

pix-art commented:
Any news on this PR?

WillSkates commented:
@garemoko I've fixed that new line there and pushed. With regards to the error checking stuff the changes are there to make sure we aren't checking the same thing twice. The previous error checking effectively made the same check as is being made in core.

@pix-art I think with this last change everything should be fine we just have to wait on @brianjmiller's approval.

brianjmiller commented:
I am working through open pull requests today and going roughly in order that they were opened, this is going to cause some conflicts as I go which I will attempt to resolve myself. Conflicts that can't be resolved in a fairly straight forward manner may have to go back to the author for correction. My intention is to work through the back log with non-PHP7 related PRs getting in first for a last minor version bump to the library, followed by the PHP7 change in a major version bump. It may not be possible to get all open PRs merged today.

WillSkates commented:
I've updated this branch with the latest changes from master.

WillSkates commented:
I know it isn't **_*strictly*_** the aim of this PR to provide 100% code coverage. Due to our conversation [here](https://github.com/RusticiSoftware/TinCanPHP/issues/54) I think it would at least give us more confidence that my changes for php7 haven't affected users running 5.5 and 5.6 as well.

@garemoko  @brianjmiller  **BIG NOTE** I'm not really proud of the way I achieved 100% cc in this case. I've tried to get it so that we can start looking at improving the CC as fast as possible. I don't for a minute think this is done even though the magic number has been achieved.

Cheers,
will

WillSkates commented:
Under PHP7 the cc report says that it has covered around 400 lines less than the 5.5 and 5.6 versions. There seems to be weird issues with PHP 7, xdebug 2.4.0 and phpunit when it comes to code coverage.  I.e. [this](https://github.com/sebastianbergmann/php-code-coverage/issues/422) and [this](https://github.com/sebastianbergmann/php-code-coverage/issues/411
).

We may have to wait a while to get a true cc report for PHP 7.

garemoko commented on tests/config/config.travis-ci.php:
```diff
@@ -2,7 +2,7 @@
 $LRSs = [
     [
-        'endpoint' => 'https://cloud.scorm.com/tc/0CKX3A0SF2/sandbox/',
+        'endpoint' => 'https://cloud.scorm.com:443/tc/0CKX3A0SF2/sandbox/',
```
Is this intentional? What's the purpose?

WillSkates replied:
Yep. The purpose is to get [this line](https://github.com/RusticiSoftware/TinCanPHP/blob/3aa60575bdca35de7667ab5dd987a7211a5e8337/src/RemoteLRS.php#L1134) to execute.

garemoko replied:
Gotcha.

brianjmiller replied:
I'm okay with that, but does that leave us open to failures when there *isn't* a port?

WillSkates commented:
Me and @garemoko had a great chat tonight and went over some things in my changes to the lib.
They are:
1. 'ignore404' option was removed in places, re adding.
2. I removed some error checking on openssl functions. I've added some line notes as to why.
3. I've made some 'else if''s 'elseif's. This does reduce the code coverage with php7 somewhat but the php7 cc report isn't accurate anyway.

@brianjmiller I know you are really busy and don't have time to give feedback, wait for my changes, give more feedback etc. Please let me know when you're free to review this and we can schedule to review it together over Skype/Zoom/GoToMeeting etc.

Cheers,
Will

WillSkates commented on src/Statement.php:
```diff
@@ -174,10 +174,8 @@ public function sign($privateKeyFile, $privateKeyPass, $options = array()) {
-                if (! openssl_x509_export($cert, $x5c, true)) {
-                    throw new \Exception('Unable to export certificate for x5c inclusion: ' . openssl_error_string());
-                }
+                
```
I've removed the check here because it seems like if $cert is valid then this line will always execute successfully.

brianjmiller replied:
But if it is invalid it returns false so we have to capture the return value of the function call and then check it, at least according to http://php.net/manual/en/function.openssl-x509-export.php?

WillSkates commented on src/Statement.php:
```diff
@@ -237,28 +303,19 @@ public function verify($options = array()) {
             if (! $cert) {
                 return array('success' => false, 'reason' => 'failed to read cert in x5c: ' . openssl_error_string());
             }
+            
```
I've removed the check here because it seems like if $cert is valid then this line will always execute successfully.

brianjmiller replied:
See question above?

WillSkates commented:
@garemoko Yep. Much better.

pix-art commented:
Do we need something else before we can merge this? :) I'd be happy to help

WillSkates commented:
@pix-art not currently, we are just waiting for @brianjmiller to look at it and give it the thumbs up.

WillSkates commented:
@brianjmiller

WillSkates commented:
Funk, apologies for accidentally closing this again.

pix-art commented:
@brianjmiller any news? :) sorry to bother again.

WillSkates commented:
@pix-art Yep. @brianjmiller is really busy working on finishing stuff for a tight deadline he has. After that we'll sit down and go through this pr. Have you tested it with your setup to see if t works?

pix-art commented:
No I haven't but i'll try that out next week when my schedule opens up a bit :) and keep you updated

WillSkates commented:
@pix-art Ah ok. Please make sure that you test against scorm cloud. @johnpbloch has encountered some issues (detailed above) that come from running the test suite against a learning locker instance.

WillSkates commented:
@brianjmiller Hows it going Brian? Any news on when we can review this together?

brianjmiller commented:
Naturally after doing all the other open PRs again this had a conflict. It was a one line change, and I think a test got broken by the other work which I'll try to take a look at. The style changes requested have never been touched and there are issues with extra new lines, new lines with just whitespace, and a general inconsistency of the style. Having said that, I'll be spending time tomorrow on this work and since it has been lingering so long and enough people seem to want it I'll be taking a look at cleaning that up. But until the style matches the rest of the lib I don't consider this mergeable, to speed this process along in the future it would be good to match the style. Until doing so, I'm not going to entertain "calls" or such, they shouldn't be necessary with the existing pool of code to match.

brianjmiller commented on src/ActivityDefinition.php:
```diff
@@ -41,13 +41,9 @@ public function __construct() {
-        ) {
+        $props = ['name', 'description', 'extensions'];
+
+        foreach ($props as $k) {
```
What is the reason for this change? Does PHP7 not handle the existing formatting?

brianjmiller commented on src/FromJSONTrait.php:
```diff
@@ -27,9 +27,10 @@ public static function fromJSON($jsonStr) {
         $called_class = get_called_class();
         return new $called_class($cfg);
     }
+
```
Change unnecessary, no reason to add blank lines.

brianjmiller commented on src/JSONParseErrorException.php:
```diff
@@ -0,0 +1,53 @@
+<?php
+/*
+    Copyright 2014 Rustici Software
```
Copyright should match the year it was added, this should be either 2015 or 2016 at this point.

brianjmiller commented on src/JSONParseErrorException.php:
```diff
@@ -0,0 +1,53 @@
+use Exception;
+
+class JSONParseErrorException extends Exception
+{
```
Bracing should match existing style.

brianjmiller commented on src/JSONParseErrorException.php:
```diff
@@ -0,0 +1,53 @@
+    public function jsonErrorMessage() {
+        return $this->jsonErrorMessage;
+    }
+
```
No blank line at bottom.

brianjmiller commented on src/LRSResponse.php:
```diff
@@ -30,4 +30,16 @@ public function __construct($success, $content, $httpResponse) {
+        $success = false;
+
+        if (($response['status'] >= 200 && $response['status'] < 300) || ($response['status'] === 404 && isset($options['ignore404']) && $options['ignore404'])) {
+                $success = true;
```
Extra indenting should be fixed.

brianjmiller commented on src/LRSResponse.php:
```diff
@@ -30,4 +30,16 @@ public function __construct($success, $content, $httpResponse) {
+                $success = true;
+        }
+        elseif ($response['status'] >= 300 && $response['status'] < 400) {
+                $content = "Unsupported status code: " . $response['status'] . " (LRS should not redirect)";
```
Prefer single quoting when not interpolating.

brianjmiller commented on src/LRSResponse.php:
```diff
@@ -30,4 +30,16 @@ public function __construct($success, $content, $httpResponse) {
         $this->httpResponse = $httpResponse;
     }
+
+    public static function fromRemoteLRSResponse(array $response, array $options, $content) {
```
Haven't used type hinting anywhere else, is there a reason to add it here? Something in PHP7 that requires it in the few locations it is included?

brianjmiller replied:
Is there a reason to move this function here (and to make it a function at all)? Does it serve a purpose in the PHP7 changes?

brianjmiller commented on src/Map.php:
```diff
@@ -41,7 +41,9 @@ public function asVersion($version = null) {
-    private function _unset($code) {
+
+    public function _unset($code)
+    {
```
What's the reason for the interface change here? If we expect to use it externally we should drop the leading '_' from the name.

brianjmiller commented on src/RemoteLRS.php:
```diff
@@ -170,12 +167,7 @@ function ($errno, $errstr, $errfile, $errline, array $errcontext) {
-            elseif ($response['status'] >= 300 && $response['status'] < 400) {
-                $content = "Unsupported status code: " . $response['status'] . " (LRS should not redirect)";
-            }
+            return LRSResponse::fromRemoteLRSResponse($response, $options, $content);
```
See question above.

brianjmiller commented on src/RemoteLRS.php:
```diff
@@ -265,40 +257,48 @@ private function _parseMultipart($boundary, $content) {
     //
     // adapted to private method, and force headers to lowercase for easy detection
+    // 
+    // Now reworked so that the variable names make a bit more sense.
```
This should be a github comment likely, temporal comments don't make sense, let the history tell the story.

brianjmiller replied:
Is all that is changing here the formatting/names of the variables? Does this change serve some other purpose and/or apply for PHP7?

brianjmiller commented on src/RemoteLRS.php:
```diff
@@ -265,40 +257,48 @@ private function _parseMultipart($boundary, $content) {
+
+            if($value === false) {
+
+                if (substr($name, 0, 1) == "\t") {
```
This is a good example of style inconsistency, 3 lines apart in the change set, in one that doesn't seem necessary at all.

brianjmiller commented on src/RemoteLRS.php:
```diff
@@ -484,42 +484,30 @@ public function retrieveVoidedStatement($id, $options = array()) {
+            'until', 'limit',
+            'format'
+        ];
+        foreach ($content as $k) {
```
Why make these separated out, and then change the original array construct for agent to be an inlined version? Does this change apply something for PHP7?

brianjmiller commented on src/RemoteLRS.php:
```diff
@@ -632,7 +620,6 @@ public function retrieveState($activity, $agent, $id) {
                 'agent'      => json_encode($agent->asVersion($this->version)),
                 'stateId'    => $id,
             ),
-            'ignore404' => true,
```
Why is ignore404 not true here?

brianjmiller commented on src/RemoteLRS.php:
```diff
@@ -831,8 +818,7 @@ public function retrieveActivityProfile($activity, $id) {
                     'activityId' => $activity->getId(),
                     'profileId'  => $id,
-                ),
-                'ignore404' => true,
```
Same for ignore404?

brianjmiller commented on src/RemoteLRS.php:
```diff
@@ -924,6 +910,7 @@ public function deleteActivityProfile($activity, $id) {
                     'activityId' => $activity->getId(),
                     'profileId'  => $id,
                 ),
+                'ignore404' => true
```
Is this from a spec change? Retrieving makes sense to ignore, but the user may want to know if they tried to delete something that they thought existed.

brianjmiller commented on src/RemoteLRS.php:
```diff
@@ -996,7 +985,7 @@ public function retrieveAgentProfile($agent, $id) {
                     'profileId' => $id,
                 ),
-                'ignore404' => true,
+                'ignore404' => true
```
Why this change? It makes it less consistent.

brianjmiller commented on src/Score.php:
```diff
@@ -100,6 +100,7 @@ public function validate($aValue) {
                 sprintf("Value must be between %s and %s", $this->min, $this->max)
             );
         }
+        return true;
```
I'm okay with this change in the major version, but ultimately it is a public interface change, the expected existing interface was any non-exception is validation. I didn't check for other validation functions, do they match this? Do we have a unit test to confirm the return value here?

brianjmiller commented on src/Statement.php:
```diff
@@ -209,6 +207,74 @@ public function sign($privateKeyFile, $privateKeyPass, $options = array()) {
+    {
+        //
+        // there is a JWS spec security issue with allowing non-RS algorithms
+        // to be specified and it is against the Tin Can spec anyways so we
```
Let's drop 'Tin Can' here, and just use 'the spec'.

brianjmiller commented on src/Statement.php:
```diff
@@ -209,6 +207,74 @@ public function sign($privateKeyFile, $privateKeyPass, $options = array()) {
         $this->addAttachment($attachment);
     }
 
+    public static function validateNameOfRSAlgorithm($algorithm)
```
Do we need this to be a separate function?

brianjmiller replied:
Can't we flip the assertion and it is 3 line code block? If anything I'd pull the algorithm name list out to make it more obvious.

brianjmiller commented on src/Statement.php:
```diff
@@ -209,6 +207,74 @@ public function sign($privateKeyFile, $privateKeyPass, $options = array()) {
+        //
+        if (in_array($algorithm, array('RS256', 'RS384', 'RS512'), true)) {
+            return true;
+        } else {
```
Don't cuddle elses.

brianjmiller commented on src/Statement.php:
```diff
@@ -209,6 +207,74 @@ public function sign($privateKeyFile, $privateKeyPass, $options = array()) {
+        }
+    }
+
+    public static function sanitizeSerializedInfo(array $serialization, array $payload, $signatureIndex)
```
Again function?

brianjmiller commented on src/Statement.php:
```diff
@@ -237,28 +303,19 @@ public function verify($options = array()) {
             $publicKeyFile = openssl_pkey_get_public($cert);
-            if (! $publicKeyFile) {
-                return array('success' => false, 'reason' => 'x5c failed to provide public key: ' . openssl_error_string());
-            }
```
? need to know if we got a valid public key file.

brianjmiller commented:
Made a closer first pass. Isolating the changes needed for a specific context would make this kind of thing much simpler in the future, IOW there seem to be relatively few changes necessary to get actual PHP7 compatibility, which could have been isolated from the test coverage pieces, which should have *definitely* been isolated from subjective code changes. I'll continue to work on this tomorrow, I want to maintain history if you feel it is important. If you'd rather take on a task of splitting this work up, and/or squashing/rebasing then feel free, let me know if you plan to.

brianjmiller commented on tests/ActivityTest.php:
```diff
@@ -44,24 +44,21 @@ public function testInstantiation() {
         $this->setExpectedException(
-            'InvalidArgumentException',
-            'Invalid JSON: ' . JSON_ERROR_SYNTAX
+            'TinCan\JSONParseErrorException'
```
While a better overall implementation this is a loss of granularity of the unit tests, we no longer test distinguishing malformed versus null/empty. We should maintain that level of detail.

brianjmiller replied:
Heh, on further inspection these tests weren't doing what they were originally thought to do anyways. `JSON_ERROR_NONE` actually ends up being '' which doesn't change the error message which is why they were passing to begin with.

brianjmiller commented on src/JSONParseErrorException.php:
```diff
@@ -0,0 +1,53 @@
+
+use Exception;
+
+class JSONParseErrorException extends Exception
```
I wonder if this should extend `InvalidArgumentException`?

brianjmiller commented on src/FromJSONTrait.php:
```diff
@@ -27,9 +27,10 @@ public static function fromJSON($jsonStr) {
         if (is_null($cfg)) {
             $err = json_last_error();
-            throw new \InvalidArgumentException("Invalid JSON: $err");
+            throw new JSONParseErrorException($cfg, $err, json_last_error_msg());
```
Based on the signature of the above exception shouldn't `$cfg` be `$jsonStr`? I think it should be the valued we tried to JSON decode, $cfg is the value as decoded (which based on the if check will always be `null`).

WillSkates commented:
Hi @brianjmiller. I'm fine to look at cleaning this up for you now that you are active here again. Closing this in favor of a 'cleaner' solution.

brianjmiller commented:
@WillSkates 👍 .

I suspect you have, but if you haven't, take a look at #50, I'm planning to accumulate the bare minimum for PHP 7 support there. It leverages the new Exception from this PR though altered a little bit and committed differently (commit message credits you though) and the corresponding changes to test files related to that change. Everything else in here I believe can be brought over in a patch (or minor) version and mostly hinges around testing AFAICT.

Splitting up the remaining parts of this PR will help get them through more quickly, for instance, straight additions of new test files (so long as they conform to the rest of the style) should be easy to add as they can't affect functionality. Followed by other adjustments to tests, and then tests and code with the latter being the most dangerous and therefore comes more scrutiny. Isolating groups of changes make it far easier to review and tweak, multiple PRs is better than larger ones.

In any case, I look forward to your submissions and appreciate the work (and patience) so far.

WillSkates commented:
Hi @brianjmiller,

Can you please set two milestones:
1. 100% code coverage. 
2. PHP7 support.

We can label my PR's against those.

Cheers,
Will

brianjmiller commented:
I think the first is a lofty goal, though you are welcome to try. If we're that close then I'd seriously look at where the metric comes from. Having said that, happy to create a milestone.

#50 achieves the second, we're there, I'm looking at the namshi/jose issue (#61) a little today. But I plan to cut a last minor version with the existing changes on master and then merge #50 and cut the major release.

If you have something that you absolutely want in a major release (because it breaks backwards compat) better get it in *very* soon.

WillSkates commented:
@brianjmiller I've already achieved it. I doubt #50 does it fully as the cc in that branch isn't sufficient. I'll see what I can get you when. If you want to push for a major version before then you can but I'd advise against it.

brianjmiller commented:
Is there some reason to advise against it? Current tests pass and they "cover" the "difficult" parts which is the JSON serialization/deserialization, and HTTP requests which are the parts likely to break in PHP 7. AFAICT all other changes in the other PR are either tests or to facilitate tests of what amounts to existing code, and therefore by virtue of it passing so should the result of #50.

WillSkates commented:
@brianjmiller I ran into places where I had to solve problems that were exposed by new tests. However I honestly can't remember them one by one so it isn't much to go on. I'll defer to you and I can work to improve your test coverage and solve them as we go in following minor versions.

Adding this to your local phpunit.xml.dist will give you a more accurate display of the code coverage as it is:

```
<log type="coverage-html" target="report" showUncoveredFiles="true" lowUpperBound="50" highLowerBound="80" />
```

brianjmiller commented:
Added, in a manor of speaking, to #50, along with some other fixes to make coverage output correctly with latest PHPUnit.

brianjmiller commented:
Heh, so I've gone through the bulk of the added files (all but one) and so far I've run into 3 issues, but none of which are related to PHP 7 support, they are just bugs that are in the 0.x series anyways, so by rights those will likely get fixed as part of it and I'll end up doing another patch release (though I hadn't planned to). I wish they'd been part of a separate PR! Oh well.

WillSkates commented:
Yeah I definitely agree that they should have been in a separate PR. Your work to split this out is very much appreciated. I'd still like to have a chat on VOIP about the lib in general if you are willing seeing as we still haven't actually met but we've both put in a fair amount of effort on this. You have my email address anyway :D.

brianjmiller commented:
A little research suggests the reason I felt "100% test coverage" was a lofty goal (and not yet achieved) and is summed up by this:

> The Opcode Coverage, Branch Coverage, and Path Coverage software metrics are not yet supported by PHP_CodeCoverage.

So the coverage is basically line coverage (which is an okay metric, but not great for regression management in a theoretical zero backwards compatibility change use case). I still think it is a worthy goal, but it is *really* important to understand just what the coverage metric is giving us and to understand the type of risks still present when making extraneous changes.

Reference: https://phpunit.de/manual/current/en/code-coverage-analysis.html
